### PR TITLE
[SDPA-Efficient] Update Alignment constraints

### PR DESCRIPTION
**Pull request by drisspg**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #140192

Fix for https://github.com/pytorch/pytorch/issues/139424
Successfully passes: https://github.com/pytorch/pytorch/issues/139424#issuecomment-2458086518
``` Shell
  File "/home/drisspg/meta/pytorch/torch/_functorch/fx_minifier.py", line 227, in minifier
    raise RuntimeError("Input graph did not fail the tester")
RuntimeError: Input graph did not fail the tester
```

### Update
Running the above inductor repro under sanitizer does not pass
```Shell

Loading inputs: 100%|██████████| 15/15 [00:00<00:00, 32.63it/s]
========= Invalid __global__ read of size 16 bytes
=========     at fmha_cutlassF_f32_aligned_64x128_rf_sm80(PyTorchMemEffAttention::AttentionKernel<float, cutlass::arch::Sm80, (bool)1, (int)64, (int)128, (int)128, (bool)1, (bool)1>::Params)+0x5940
=========     by thread (30,1,0) in block (127,0,1)
=========     Address 0x7f38582f8400 is out of bounds
=========     and is 1 bytes after the nearest allocation at 0x7f3858278400 of size 524,288 bytes
=========     Saved host backtrace up to driver entry point at kernel launch time
=========     Host Frame: [0x2f16cf]
=========                in /lib64/libcuda.so.1
=========     Host Frame: [0x15803]
=========                in /usr/local/cuda-12.4/lib64/libcudart.so.12
=========     Host Frame:cudaLaunchKernel [0x75230]
=========                in /usr/local/cuda-12.4/lib64/libcudart.so.12
=========     Host Frame:cudaError cudaLaunchKernel<char>(char const*, dim3, dim3, void**, unsigned long, CUstream_st*) in /usr/local/cuda-12.4/include/cuda_runtime.h:216 [0x6f73d16]
=========                in /home/drisspg/meta/pytorch/torch/lib/libtorch_cuda.so
=========     Host Frame:__device_stub__Z40fmha_cutlassF_f32_aligned_64x128_rf_sm80N22PyTorchMemEffAttention15AttentionKernelIfN7cutlass4arch4Sm80ELb1ELi64ELi128ELi128ELb1ELb1EE6ParamsE(PyTorchMemEffAttention::AttentionKernel<float, cutlass::arch::Sm80, true, 64, 128, 128, true, true>::Params&) in /tmp/tmpxft_001f819e_00000000-6_cutlassF_f32_aligned.cudafe1.stub.c:1 [0x6f7328d]
=========                in /home/drisspg/meta/pytorch/torch/lib/libtorch_cuda.so
=========     Host Frame:fmha_cutlassF_f32_aligned_64x128_rf_sm80(PyTorchMemEffAttention::AttentionKernel<float, cutlass::arch::Sm80, true, 64, 128, 128, true, true>::Params) in /home/drisspg/meta/pytorch/aten/src/ATen/native/transformers/cuda/mem_eff_attention/kernels/cutlassF_f32_aligned.cu:147 [0x6f732a0]
=========                in /home/drisspg/meta/pytorch/torch/lib/libtorch_cuda.so
=========     Host Frame:auto at::native::_efficient_attention_forward(at::Tensor const&, at::Tensor const&
````

Running with
```Shell
CUDA_LAUNCH_BLOCKING=1 PYTORCH_NO_CUDA_MEMORY_CACHING=1 compute-sanitizer --tool memcheck pytest test/test_transformers.py -sk test_mem_eff_stride_zero_aligned
```
Screenshot 2024-11-08 at 4 16 27 PM

### CODE REVIEW

**Comment by Skylion007** [comment]
Seems like a good backport candidate.

File: test/test_transformers.py [hunk 1]
```diff
@@ -1912,6 +1912,47 @@ def test_flash_attention_fail_with_non_square_causal_attention(self, device):
...
+            (16384, 0, 2, 0),
+            (8192, 1, 0, 0),
+            (0, 8192, 1, 0),
+            (8192, 0, 0, 1),
+            (0, 16384, 2, 0),
+            (131072, 0, 16, 0),
+            (32768, 16384, 8, 1)
+        ],
+    )
+    def test_mem_eff_stride_zero_aligned(self, device, strides: Tuple[int]):
```
Comment (Skylion007) [inline test/test_transformers.py:1934]:
```suggestion
    def test_mem_eff_stride_zero_aligned(self, device, strides: Tuple[int, ...]):
```

**Comment by ringohoffman** [comment]
Any update here?

**Comment by drisspg** [comment]
@ringohoffman We have had some internal planning and didn't have much chance to work on it this week. We'll pick it back up next week.

**Review by eellison — COMMENTED** [review]
how do you want to move forward with this ?

**Comment by drisspg** [comment]
cc @eellison Did some quick debugging and looking at the repros I have in unit tests:

My past unit tests were wrong

size 					:torch.Size([2, 32, 8192, 8192])
strides				: (65536, 0, 8, 0)
offset 				: 0
storage_size	: 196608

And in the inductor repro:
size                  : torch.Size([2, 32, 8192, 8192])
strides             : (65536, 0, 8, 0)
offset				: 0 
storage_size   : 131072

updated to get the correct number of numels of storage.

I think the code that prevents us from making this change is here: https://github.com/pytorch/pytorch/blob/851edf208b308789087f263de739607ea25c8a0e/aten/src/ATen/native/transformers/cuda/mem_eff_attention/kernel_forward.h#L809

We would need to update the local layout w/ the 0 striding for the n dim

**Comment by eellison** [comment]
@ringohoffman yes there will be. ran into https://github.com/pytorch/pytorch/issues/141888 while fixing it, once that is fixed will submit pr for htis

**Pull request closed without merging**